### ISSUE TITLE
Display of doubly accented Thai characters

### ISSUE THREAD

**Issue by salvipeter** [vim/vim#13043]
### Steps to reproduce

1. Use a Thai-capable font (e..g the fixed-width Tlwg Typist)
2. Paste น้ำ นี้
3. Alternatively, load this keymap: http://salvi.chaosnet.org/snippets/thai_utf-8.vim
4. ... and type "ohe ouh"

### Expected behaviour

The words should look like the left-hand sides of the equations in this image:
![thai-emacs](https://github.com/vim/vim/assets/104294/8e033c9f-36a5-4aa9-8e80-13b9f134ad6e)

Instead they look like this:
![thai-gvim](https://github.com/vim/vim/assets/104294/014bb553-6701-46ec-a6d0-b9f431293d8a)

(Using vim in a terminal with a Thai font leads to similar results.)

### Version of Vim

9.0.1736

### Environment

OS: Manjaro (6.0.11 kernel)
Terminal: xfce4-terminal 1.1.0-1 (or gvim)
$TERM: xterm-256color
shell: bash 5.1.016

**Comment by ychin** [comment]
For using in terminal, I think that's the bug with your terminal. E.g. I used it in Ubuntu's terminal and it works fine. I did reproduce this in GTK gVim though, so I think that's a bug with the GTK GUI.

**Comment by salvipeter** [comment]
@ychin, you are right, in the case of using vim in a terminal, it seems to be the terminal's problem. Exactly what terminal did you use? I have tried xfce-terminal, xterm, rxvt-unicode, lxterminal, gnome-terminal, konsole, terminology, terminator, guake and st, with several TLWG fonts (Tlwg Typist, Tlwg Mono, Tlwg Typo, Tlwg Typewriter), none of which produced the correct output. (In many cases, the second example is displayed correctly, but not the first - note the correct position of the small circle.)
Using the Emacs Gtk GUI with any of the above fonts works well (e.g. with: -PfEd-Tlwg Typist-regular-normal-normal-*-13-*-*-*-*-0-iso10646-1), but fails when running Emacs in a terminal, of course.

**Comment by ychin** [comment]
Right , actually in terminal it does look wrong. The circle is overlapped. Seems like the existence of ำ affects the "น้", but the logic isn't handling that.
